bob-access/src/authenticator/basic.rs

```diff
@@ -1,65 +1,136 @@
+                }
+            }
+            let mut nodes_creds = self.nodes.write().expect("nodes credentials lock");
+            nodes_creds.extend(resolved.drain());
             Ok(())
         } else {
             let message = "nodes credentials missing ip or node name";
             Err(Error::CredentialsNotProvided(message.to_string()))
         }
     }
 
-    fn check_node_request(&self, node_name: &String, ip: Option<IpAddr>) -> Option<bool> {
-        if self.nodes.is_empty() {
-            warn!("nodes credentials not set");
+    fn spawn_resolver(&self, cred: DeclaredCredentials) {
+        tokio::spawn(Self::resolve_worker(self.nodes.clone(), cred, self.resolve_sleep_dur_ms));
+    }
+
+    async fn resolve_worker(nodes: Arc<RwLock<NodesCredentials>>, cred: DeclaredCredentials, sleep_dur_ms: u64) {
+        let hostname = cred.hostname().as_ref().expect("resolve worker without hostname");
+        let mut addr: Option<Vec<SocketAddr>> = None;
+        let mut cur_sleep_dur_ms = 100;
+        while addr.is_none() || (addr.is_some() && addr.as_ref().unwrap().len() == 0) {
+            tokio::time::sleep(Duration::from_millis(cur_sleep_dur_ms)).await;
+
+            addr = match lookup_host(hostname).await {
+                Ok(address) => Some(address.collect()),
+                _ => None
+            };
+
+            cur_sleep_dur_ms = sleep_dur_ms.min(cur_sleep_dur_ms * 2);
         }
-        self.nodes
-            .get(ip.as_ref()?)
-            .map(|creds| {
-                for cred in creds {
-                    if let Some(CredentialsKind::InterNode(other_name)) = cred.kind() {
-                        if node_name == other_name {
-                            return true;
-                        }
+
+        let addr = addr.expect("somehow addr is none");
+        let mut nodes = nodes.write().expect("nodes credentials lock");
+        if let Some(CredentialsKind::InterNode(nodename)) = cred.kind() {
+            if let Some(creds) = nodes.get_mut(nodename) {
+                creds.add_addresses(addr);
+                creds.set_resolved();
+            }
+        } else {
+            error!("resolved credentials are not internode");
+        }
+    }
+
+    fn check_resolve(&self, nodename: &str, unresolved: bool) {
+        let mut nodes = self.nodes.write().expect("nodes credentials lock");
+        if let Some(node) = nodes.get_mut(nodename) {
+            if node.needs_resolve(unresolved) {
+                node.set_pending();
+                self.spawn_resolver(node.clone());
+            }
+        }
+    }
+
+    fn check_node_request(&self, node_name: &String, ip: Option<SocketAddr>) -> bool {
+        let mut unresolved = false;
+        {
+            if ip.is_none() {
+                return false;
+            }
+            let ip = ip.unwrap().ip();
+            let nodes = self.nodes.read().expect("nodes credentials lock");
+            if nodes.is_empty() {
```

Comment:
This check is not needed. Below you call `nodes.get(node_name)` that will return `None` for the empty list